### PR TITLE
Pundit 2/3: Convert all models to use Pundit for scopes

### PR DESCRIPTION
**Pull request by marten**
Converted all remaining models over to Pundit policy scopes.

Main changes:

* All models now use Pundit for their scoping.
* All controllers now use Pundit, so LegacyPolicy is removed.
* Removed all `RoleControl`
* Some controllers like the classifications controller had scopes that did filtereing based on params. Those used to be passed all the way into the model layer so that the model could then pick things out of the params. This is much cleaner now: the Scope returns *all* that could possibly be retrieved (as an AR Relation, of course), and the controller's task is to then do filtering on that based on params.
* ApiUser is no longer an Actor. All policy lookups go through Pundit, not through ApiUser. This means that ApiUser is merely a wrapper for User that adds some logic around supporting non-logged-in as an actual object instead of just `nil`, and adds admin-flaggyness.
* All scopes are now tested at the Policy level. This means that a next PR will go through all the controller tests and simplify them, since there is no need to test various types of users through the controller layer.

Notes:

* The PunditInterop module is used to support `can_be_linked` which will be the target of the next PR in this series. In that one it will be removed.
* The old can_by_role scopes used to take into account a `project_editor` and `collection_editor`. Neither of these exist in the database, and therefore these are no longer included now.
* I will also be revisiting the idea of `PunditPolicy` as a module, and how `controlled_resources` works / where it lives as a method.

# Review checklist

- [ ] First, the most important one: is this PR small enough that you can actually review it? Feel free to just reject a branch if the changes are hard to review due to the length of the diff.
- [ ] If there are any migrations, will they the previous version of the app work correctly after they've been run (e.g. the don't remove columns still known about by ActiveRecord).
- [ ] If anything changed with regards to the public API, are those changes also documented in the `apiary.apib` file?
- [ ] Are all the changes covered by tests? Think about any possible edge cases that might be left untested.

### CODE REVIEW

File: spec/policies/organization_policy_spec.rb [hunk 1]
```diff
@@ -9,172 +9,152 @@
...
-        it 'should not include private resources' do
-          unlisted_organization.save!
-          expect(resolved_scope).not_to include(unlisted_organization)
-        end
-      end
-
-      context 'for the resource owner' do
-        let(:api_user) { ApiUser.new(resource_owner) }
+    before do
+      listed_organization.save!
```
[the remark this answers is not in the document]
Comment (marten) [reply spec/policies/organization_policy_spec.rb]: They need to be in the db before the `is_expected` is called. I guess there's basically three possible patterns I can think of:

```
let!(:listed_organization) { create :organization }
```

or 

```
let(:organization) { create :organization }
before { organization }
```

or 

```
let(:organization) { build :organization }
before { organization.save! }
```

I always seem to miss the exclamation mark in `let!` and the second one looks "funny" to me.

File: app/policies/user_group_policy.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
+class UserGroupPolicy < ApplicationPolicy
+  # The old can_by_role scopes used to take into accuont a `project_editor` and `collection_editor`.
```
[the remark this answers is not in the document]
Comment (marten) [reply app/policies/user_group_policy.rb]: Agreed. I left it mostly as a note to self while coding. I'll put it in the PR body.

**Comment by marten** [comment]
It's good to hear that both @zwolf and you got confused about the magic of where the Policy comes from. I'll try to improve on that, even if it's Pundit's own magic

File: spec/lib/http_cacheable_spec.rb [hunk 3]
```diff
@@ -50,7 +50,7 @@ def make_private(public_resource)
         end
 
         it "should return false with a private resource" do
-          parent_relation = resource_klass.parent_relation
+          parent_relation = :project
```
[the remark this answers is not in the document]
Comment (marten) [reply spec/lib/http_cacheable_spec.rb]: I don't think I follow

[the remark this answers is not in the document]
Comment (camallen) [reply spec/lib/http_cacheable_spec.rb]: removing `resource_klass.parent_relation` from `parental_controlled.rb` changes the way the `http_cacheable` tests which resource to lookup for private resources, https://github.com/zooniverse/Panoptes/blob/aeb660828a0b9db92e84e53b50bd967e8421c006/lib/http_cacheable.rb#L60

I think the new behaviour still works but i'm not sure the tests here cover it or test code that isn't needed anymore.

[the remark this answers is not in the document]
Comment (marten) [reply spec/lib/http_cacheable_spec.rb]: Just noting here that this is resolved

File: app/controllers/concerns/recents.rb [hunk 4]
```diff
@@ -1,4 +1,6 @@
 module Recents
+  extend ActiveSupport::Concern
```
[the remark this answers is not in the document]
Comment (marten) [reply app/controllers/concerns/recents.rb]: Not anymore, just somewhere halfway during the refactoring. Will remove it again

**Comment by marten** [comment]
As for deploying, I think there's two options

* merge and deploy one every day/few days
* merge everything and deploy all at once

I don't have a strong preference either way. You?

**Comment by camallen** [comment]
Not real plan, if you are happy that you can isolate any bugs / errors in the changes via error reports / code knowledge then the big one will probably be easier.

**Pull request merged**